### PR TITLE
🖨️ Debug Printing Improvements

### PR DESCRIPTION
**Pull request by TOTBWF**
## Patch Description

This PR adds a new `Debug` module, which consists of some simple debugging printers that can be toggled on and off at will. It also adds names to tactics, which allow us to trace what tactics are being executed for easier debugging, along with a new `--debug` flag for enabling debug printing.

### CODE REVIEW

File: src/core/Refiner.ml [hunk 1]
```diff
@@ -99,7 +99,7 @@ struct
     Format.fprintf fmt "Emitted hole:@,  @[<v>%a@]@." (S.pp_partial_sequent bdry_sat ctx) (tm, stp)
 
   let probe_chk name tac =
-    T.Chk.brule @@ fun (tp, phi, clo) ->
+    T.Chk.brule ~name:"probe_chk" @@ fun (tp, phi, clo) ->
```
Comment (favonia) [inline src/core/Refiner.ml:102]:
```suggestion
    T.Chk.brule ~name:"⬇️" @@ fun (tp, phi, clo) ->
```

File: src/core/Refiner.ml [hunk 2]
```diff
@@ -108,13 +108,13 @@ struct
     s
 
   let probe_boundary probe tac =
-    T.Chk.brule @@ fun (tp, phi, clo) ->
+    T.Chk.brule ~name:"probe_boundary" @@ fun (tp, phi, clo) ->
```
Comment (favonia) [inline src/core/Refiner.ml:111]:
```suggestion
    T.Chk.brule ~name:"⬇️🪟" @@ fun (tp, phi, clo) ->
```

File: src/core/Refiner.ml [hunk 3]
```diff
@@ -108,13 +108,13 @@ struct
...
     let* probe_tm = T.Chk.run probe tp in
     let* () = print_boundary probe_tm tp phi clo in
     T.Chk.brun tac (tp, phi, clo)
 
   let probe_syn name tac =
-    T.Syn.rule @@
+    T.Syn.rule ~name:"probe_syn" @@
```
Comment (favonia) [inline src/core/Refiner.ml:117]:
```suggestion
    T.Syn.rule ~name:"⬆️" @@
```

File: src/core/Refiner.ml [hunk 4]
```diff
@@ -160,19 +160,19 @@ struct
     RM.ret (D.UnstableCut (cut, D.KSubOut (phi, clo)), [])
 
   let silent_hole name : T.Chk.tac =
-    T.Chk.brule @@ fun (tp, phi, clo) ->
+    T.Chk.brule ~name:"silent_hole" @@ fun (tp, phi, clo) ->
```
Comment (favonia) [inline src/core/Refiner.ml:163]:
```suggestion
    T.Chk.brule ~name:"⬇️🔇" @@ fun (tp, phi, clo) ->
```

File: src/core/Refiner.ml [hunk 5]
```diff
@@ -160,19 +160,19 @@ struct
...
     let* cut = make_hole name (tp, phi, clo) in
     RM.quote_cut cut
 
   let unleash_hole name : T.Chk.tac =
     Probe.probe_chk name @@
-    T.Chk.brule @@ fun (tp, phi, clo) ->
+    T.Chk.brule ~name:"unleash_hole" @@ fun (tp, phi, clo) ->
```
Comment (favonia) [inline src/core/Refiner.ml:169]:
```suggestion
    T.Chk.brule ~name:"⬇️⛳" @@ fun (tp, phi, clo) ->
```

File: src/core/Refiner.ml [hunk 6]
```diff
@@ -160,19 +160,19 @@ struct
...
     let* cut = make_hole name (tp, phi, clo) in
     RM.quote_cut cut
 
   let unleash_syn_hole name : T.Syn.tac =
     Probe.probe_syn name @@
-    T.Syn.rule @@
+    T.Syn.rule ~name:"unleash_syn_hole" @@
```
Comment (favonia) [inline src/core/Refiner.ml:175]:
```suggestion
    T.Syn.rule ~name:"⬆️⛳" @@
```

File: src/core/Refiner.ml [hunk 7]
```diff
@@ -183,7 +183,7 @@ end
 module Sub =
 struct
   let formation (tac_base : T.Tp.tac) (tac_phi : T.Chk.tac) (tac_tm : T.var -> T.Chk.tac) : T.Tp.tac =
-    T.Tp.rule @@
+    T.Tp.rule ~name:"Sub.formation" @@
```
Comment (favonia) [inline src/core/Refiner.ml:186]:
```suggestion
    T.Tp.rule ~name:"🌗" @@
```

File: src/core/Refiner.ml [hunk 8]
```diff
@@ -195,7 +195,7 @@ struct
     S.Sub (base, phi, tm)
 
   let intro (tac : T.Chk.tac) : T.Chk.tac =
-    T.Chk.brule @@
+    T.Chk.brule ~name:"Sub.intro" @@
```
Comment (favonia) [inline src/core/Refiner.ml:198]:
```suggestion
    T.Chk.brule ~name:"🌗📥" @@
```

File: src/core/Refiner.ml [hunk 9]
```diff
@@ -216,7 +216,7 @@ struct
       RM.expected_connective `Sub tp
 
   let elim (tac : T.Syn.tac) : T.Syn.tac =
-    T.Syn.rule @@
+    T.Syn.rule ~name:"Sub.elim" @@
```
Comment (favonia) [inline src/core/Refiner.ml:219]:
```suggestion
    T.Syn.rule ~name:"🌗📤" @@
```

File: src/core/Refiner.ml [hunk 10]
```diff
@@ -419,7 +419,7 @@ struct
       RM.expected_connective `LockedPrf tp
 
   let unlock prf bdy =
-    T.Chk.rule @@
+    T.Chk.rule ~name:"LockedPrf.unlock" @@
```
Comment (favonia) [inline src/core/Refiner.ml:422]:
```suggestion
    T.Chk.rule ~name:"🔓" @@
```

File: src/core/Refiner.ml [hunk 11]
```diff
@@ -405,12 +405,12 @@ end
...
 struct
   let formation tac_phi =
-    T.Tp.rule @@
+    T.Tp.rule ~name:"LockedPrf.formation" @@
     let+ phi = T.Chk.run tac_phi D.TpCof in
     S.TpLockedPrf phi
 
   let intro =
-    T.Chk.rule @@
+    T.Chk.rule ~name:"LockedProf.intro" @@
```
Comment (favonia) [inline src/core/Refiner.ml:413]:
```suggestion
    T.Chk.rule ~name:"🔒" @@
```

File: src/core/Refiner.ml [hunk 12]
```diff
@@ -405,12 +405,12 @@ end
 module LockedPrf =
 struct
   let formation tac_phi =
-    T.Tp.rule @@
+    T.Tp.rule ~name:"LockedPrf.formation" @@
```
Comment (favonia) [inline src/core/Refiner.ml:408]:
```suggestion
    T.Tp.rule ~name:"🔐" @@
```

File: src/core/Refiner.ml [hunk 13]
```diff
@@ -228,19 +228,19 @@ end
 module Dim =
 struct
   let formation : T.Tp.tac =
-    T.Tp.virtual_rule @@
+    T.Tp.virtual_rule ~name:"Dim.formation" @@
```
Comment (favonia) [inline src/core/Refiner.ml:231]:
```suggestion
    T.Tp.virtual_rule ~name:"🔟" @@
```

File: src/core/Refiner.ml [hunk 14]
```diff
@@ -228,19 +228,19 @@ end
...
     RM.ret S.TpDim
 
   let dim0 : T.Chk.tac =
-    T.Chk.rule @@
+    T.Chk.rule ~name:"Dim.dim0" @@
```
Comment (favonia) [inline src/core/Refiner.ml:235]:
```suggestion
    T.Chk.rule ~name:"0️⃣" @@
```

File: src/core/Refiner.ml [hunk 15]
```diff
@@ -228,19 +228,19 @@ end
...
     function
     | D.TpDim ->
       RM.ret S.Dim0
     | tp ->
       RM.expected_connective `Dim tp
 
   let dim1 : T.Chk.tac =
-    T.Chk.rule @@
+    T.Chk.rule ~name:"Dim.dim1" @@
```
Comment (favonia) [inline src/core/Refiner.ml:243]:
```suggestion
    T.Chk.rule ~name:"1️⃣" @@
```

File: src/core/Refiner.ml [hunk 16]
```diff
@@ -252,21 +252,21 @@ struct
     | 0 -> dim0
     | 1 -> dim1
     | n ->
-      T.Chk.rule @@ fun _ ->
+      T.Chk.rule ~name:"Dim.literal" @@ fun _ ->
```
Comment (favonia) [inline src/core/Refiner.ml:255]:
```suggestion
      T.Chk.rule ~name:"🔢" @@ fun _ ->
```

File: src/core/Refiner.ml [hunk 17]
```diff
@@ -285,7 +285,7 @@ struct
       expected_cof tp
 
   let meet tacs =
-    T.Chk.rule @@
+    T.Chk.rule ~name:"Cof.meet" @@
```
Comment (favonia) [inline src/core/Refiner.ml:288]:
```suggestion
    T.Chk.rule ~name:"🥅✖️" @@
```

File: src/core/Refiner.ml [hunk 18]
```diff
@@ -276,7 +276,7 @@ struct
       expected_cof tp
 
   let join tacs =
-    T.Chk.rule @@
+    T.Chk.rule ~name:"Cof.join" @@
```
Comment (favonia) [inline src/core/Refiner.ml:279]:
```suggestion
    T.Chk.rule ~name:"🥅➕" @@
```

File: src/core/Refiner.ml [hunk 19]
```diff
@@ -252,21 +252,21 @@ struct
...
-    T.Tp.virtual_rule @@
+    T.Tp.virtual_rule ~name:"Cof.formation" @@
     RM.ret S.TpCof
 
   let expected_cof =
     RM.expected_connective `Cof
 
   let eq tac0 tac1 =
-    T.Chk.rule @@
+    T.Chk.rule ~name:"Cof.eq" @@
```
Comment (favonia) [inline src/core/Refiner.ml:269]:
```suggestion
    T.Chk.rule ~name:"🥅↔️" @@
```

File: src/core/Refiner.ml [hunk 20]
```diff
@@ -252,21 +252,21 @@ struct
...
-      T.Chk.rule @@ fun _ ->
+      T.Chk.rule ~name:"Dim.literal" @@ fun _ ->
       RM.refine_err @@ Err.ExpectedDimensionLiteral n
 end
 
 module Cof =
 struct
   let formation : T.Tp.tac =
-    T.Tp.virtual_rule @@
+    T.Tp.virtual_rule ~name:"Cof.formation" @@
```
Comment (favonia) [inline src/core/Refiner.ml:262]:
```suggestion
    T.Tp.virtual_rule ~name:"🥅" @@
```

File: src/core/Refiner.ml [hunk 21]
```diff
@@ -352,7 +352,7 @@ struct
     end
 
     let split (branches : branch_tac list) : T.Chk.tac =
-      T.Chk.brule @@ fun (tp, psi, psi_clo) ->
+      T.Chk.brule ~name:"Split.split" @@ fun (tp, psi, psi_clo) ->
```
Comment (favonia) [inline src/core/Refiner.ml:355]:
```suggestion
      T.Chk.brule ~name:"🪚" @@ fun (tp, psi, psi_clo) ->
```

File: src/core/Refiner.ml [hunk 22]
```diff
@@ -388,12 +388,12 @@ end
 module Prf =
 struct
   let formation tac_phi =
-    T.Tp.virtual_rule @@
+    T.Tp.virtual_rule ~name:"Prf.formation" @@
```
Comment (favonia) [inline src/core/Refiner.ml:391]:
```suggestion
    T.Tp.virtual_rule ~name:"🛂" @@
```

File: src/core/Refiner.ml [hunk 23]
```diff
@@ -440,14 +440,14 @@ module Pi =
 struct
   let formation : (T.Tp.tac, T.Tp.tac) quantifier =
     fun tac_base (nm, tac_fam) ->
-      T.Tp.rule @@
+      T.Tp.rule ~name:"Pi.formation" @@
```
Comment (favonia) [inline src/core/Refiner.ml:443]:
```suggestion
      T.Tp.rule ~name:"🥧" @@
```

File: src/core/Refiner.ml [hunk 24]
```diff
@@ -440,14 +440,14 @@ module Pi =
...
       let* base = T.Tp.run_virtual tac_base in
       let* vbase = RM.lift_ev @@ Sem.eval_tp base in
       let+ fam = T.abstract ~ident:nm vbase @@ fun var -> T.Tp.run @@ tac_fam var in
       S.Pi (base, nm, fam)
 
   let intro ?(ident = `Anon) (tac_body : T.var -> T.Chk.tac) : T.Chk.tac =
-    T.Chk.brule @@
+    T.Chk.brule ~name:"Pi.intro" @@
```
Comment (favonia) [inline src/core/Refiner.ml:450]:
```suggestion
    T.Chk.brule ~name:"🥣" @@
```

File: src/core/Refiner.ml [hunk 25]
```diff
@@ -458,7 +458,7 @@ struct
       RM.expected_connective `Pi tp
 
   let apply tac_fun tac_arg : T.Syn.tac =
-    T.Syn.rule @@
+    T.Syn.rule ~name:"Pi.apply" @@
```
Comment (favonia) [inline src/core/Refiner.ml:461]:
```suggestion
    T.Syn.rule ~name:"🍲" @@
```

File: src/core/Refiner.ml [hunk 26]
```diff
@@ -508,7 +508,7 @@ struct
 
 
   let pi2 tac : T.Syn.tac =
-    T.Syn.rule @@
+    T.Syn.rule ~name:"Sg.pi2" @@
```
Comment (favonia) [inline src/core/Refiner.ml:511]:
```suggestion
    T.Syn.rule ~name:"🥈" @@
```

File: src/core/Refiner.ml [hunk 27]
```diff
@@ -498,7 +498,7 @@ struct
       RM.expected_connective `Sg tp
 
   let pi1 tac : T.Syn.tac =
-    T.Syn.rule @@
+    T.Syn.rule ~name:"Sg.pi1" @@
```
Comment (favonia) [inline src/core/Refiner.ml:501]:
```suggestion
    T.Syn.rule ~name:"🥇" @@
```

File: src/core/Refiner.ml [hunk 28]
```diff
@@ -477,14 +477,14 @@ module Sg =
...
-      T.Tp.rule @@
+      T.Tp.rule ~name:"Sg.formation" @@
       let* base = T.Tp.run tac_base in
       let* vbase = RM.lift_ev @@ Sem.eval_tp base in
       let+ fam = T.abstract ~ident:nm vbase @@ fun var -> T.Tp.run @@ tac_fam var in
       S.Sg (base, nm, fam)
 
   let intro (tac_fst : T.Chk.tac) (tac_snd : T.Chk.tac) : T.Chk.tac =
-    T.Chk.brule @@
+    T.Chk.brule ~name:"Sg.intro" @@
```
Comment (favonia) [inline src/core/Refiner.ml:487]:
```suggestion
      T.Tp.rule ~name:"🥢" @@
```

File: src/core/Refiner.ml [hunk 29]
```diff
@@ -477,14 +477,14 @@ module Sg =
 struct
   let formation : (T.Tp.tac, T.Tp.tac) quantifier =
     fun tac_base (nm, tac_fam) ->
-      T.Tp.rule @@
+      T.Tp.rule ~name:"Sg.formation" @@
```
Comment (favonia) [inline src/core/Refiner.ml:480]:
```suggestion
      T.Tp.rule ~name:"🥡" @@
```

**Comment by favonia** [comment]
This is so much fun but let me stop now before it goes out of control... :sweat_smile:

**Review by favonia — APPROVED** [review]

**Comment by favonia** [comment]
In the long run, we could consider the approaches based on `ppx` to massage the backtraces.

**Review by jonsterling — APPROVED** [review]
Great idea! Looks good.

**Pull request merged**